src/api/client-event-manager-api.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright 2019 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {AnalyticsEvent,EventOriginator} from '../proto/api_messages';
+
+/** @enum {number}  */
+export const FilterResult = {
+  /** the event is allowed to proceed to the listeners */
+  PROCESS_EVENT: 0,
+  /** the event is canceled and the listeners are not informed about it */
+  CANCEL_EVENT: 1,
+};
+
+/**
+ * Defines a client event in SwG
+ * Properties:
+ * - eventType: Required. The AnalyticsEvent type that occurred.
+ * - eventOriginator: Required.  The codebase that initiated the event.
+ * - additionalParameters: Optional.  A JSON object to store generic data.
+ * - isFromUserAction: Optional.  True if the user took an action to generate
+ *   the event.
+ *
+ *  @typedef {{
+ *    eventType: !AnalyticsEvent,
+ *    eventOriginator: !EventOriginator,
+ *    isFromUserAction: ?boolean,
+ *    additionalParameters: ?Object,
+ * }}
+ */
+export let ClientEvent;
+
+/**
+ * @interface
+ */
+export class ClientEventManagerApi {
+  /**
+   * Call this function to log an event. The registered listeners will be
+   * invoked unless the event is filtered.
+   * @param {!function(!ClientEvent)} callback
+   */
+  registerEventListener(callback) { }
+
+  /**
+   * Register a filterer for events if you need to potentially prevent the
+   * listeners from hearing about it.  A filterer should return
+   * FilterResult.CANCEL_EVENT to prevent listeners from hearing about the
+   * event.
+   * @param {!function(!ClientEvent):FilterResult} callback
+   */
+  registerEventFilterer(callback) { }
+
+  /**
+   * Call this function to log an event.  The registered listeners will be
+   * invoked unless the event is filtered out.  Returns a promise to call the
```

Comment:
Can you clarify? I'm a bit sticky about the return type because typically logging is a fire-and-forget operation. What do you mean here? Who should use the promise response and when? Can two logEvent be called one after the other w/o waiting for the other to complete? Does it yield when listeners process the event? What does "unless it is filtered out" means? Will the promise not yield in that case?

Reply:
The idea was that the testing code needed to wait for the asynchronous logging calls to finish for it to verify the listeners were actually called.  Sohani convinced me nothing else would ever care and so a private lastAction_ variable in the event manager is sufficient for this since it will only ever be used by testing code.

Yes two logging events could be called simultaneously (that was why I wanted a promise instead of a private variable that would only track the most recent call). 

Unless it is filtered out means a filterer decided to cancel the event.  Which would yield the promise but not actually call the listeners.

Reply:
Yeah. Something like `lastAction_` or `lastOperation_` could help here. While it's preferred that the API itself is fully testable, sometimes that doesn't work and we don't want test code to unduly influence the API too.

If you're not comfortable accessing private vars in the tests, sometimes what I do is create a plain export function for testing. E.g.

```
class MyApi {
  operation() {
    this.lastAction_ = new Promise(...);
  }
}

/**
 * ...
 */
export function getLastActionForTesting(instance) {
  return instance.lastAction_;
}
```

Note that any such method suffixed with "ForTesting" is removed by our compiler. This makes it obvious that this exists only for testing and no PROD code can actually rely on it.

Reply:
I don't mind accessing the private variable in the testing code as long as it's only the testing code that is in client-event-manager-test.js.  If for some reason it needed to be accessed in other test code somewhere else I might prefer it to have a public method.

Reply:
Good to know about the ForTesting suffix though!